processor/fanoutconnector.go

```diff
@@ -57,10 +81,142 @@ var _ TraceProcessor = (*traceFanOutConnector)(nil)
 // ConsumeTraceData exports the span data to all trace consumers wrapped by the current one.
 func (tfc traceFanOutConnector) ConsumeTraceData(ctx context.Context, td consumerdata.TraceData) error {
 	var errs []error
-	for _, tc := range tfc {
-		if err := tc.ConsumeTraceData(ctx, td); err != nil {
+
+	// Fan out to first len-1 consumers.
+	for i := 0; i < len(tfc)-1; i++ {
+		// Create a clone of data. We need to clone because consumers may modify the data.
+		clone, err := cloneTraceData(&td)
+		if err != nil {
 			errs = append(errs, err)
+			break
+		} else {
+			if err := tfc[i].ConsumeTraceData(ctx, *clone); err != nil {
+				errs = append(errs, err)
+			}
 		}
 	}
+
+	if len(tfc) > 0 {
+		// Give the original data to the last consumer.
+		lastTc := tfc[len(tfc)-1]
+		if err := lastTc.ConsumeTraceData(ctx, td); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
 	return oterr.CombineErrors(errs)
 }
+
+func cloneTraceData(td *consumerdata.TraceData) (*consumerdata.TraceData, error) {
```

Comment:
Consider `proto.Clone`: https://github.com/golang/protobuf/blob/master/proto/clone.go.

Reply:
Also by using `proto.Clone` this func doesn't need to return an error. Same for cloneMetricsData.

Reply:
Thanks, I didn't know there is `Clone`, this is very useful!

Reply:
Good suggestion @songy23 this should simplify the code.

Reply:
Can we benchmark the Clone? I am curios how much it costs.

Reply:
Added Benchmark100SpanClone.

Cloning one span takes about 8 microseconds on my machine. Not cheap but not extremely expensive (this should be about 8% of total cost based on 10K spans/sec figure per CPU that we currently have).